### DISCUSSION TITLE
support app passwords for legacy applications

### DISCUSSION THREAD

**Discussion by ibotty** [zitadel/zitadel#3707] [Ideas]
Say I have deployed dovecot with oauth2 authentication.  Now legacy clients like thunderbird can't connect.  It would be great to have an integrated way in zitadel to have app passwords.  These are application- and device-specific passwords.

This has (imo) two parts:
[ ] storing the app passwords in the db and having an API that checks the password and returns a valid token (or just a `true` or `false`),
[ ] have some UI to manage the app passwords.

**Comment by fforootd** [comment]
Hi @ibotty 

Interesting thought. Quick question who would consume the secrets in that case from ZITADEL?

Do you think an identity aware proxy would call the api when presented with a token and call upstream to the target service with the secret? Or did you think of something else.

**Reply by ibotty** [reply]
I am not sure I understand the question.

What I did implement for one deployment was a hook in dovecot that asked the database if the given (clear text password) was right.  The query matched against client-IP(-range), app name (`imap` in this case) and the hash of the password.  It could also use the [CheckPassword Interface](https://doc.dovecot.org/configuration_manual/authentication/checkpassword/#authentication-checkpassword) to call a shell script that calls a web api.

The important point is to have a way to check (preferably via HTTP-API) whether a given password is one of the valid passwords.  Ideally in constant time and protected by an oauth token.  All that could be part of the script (or binary program) though.

**Reply by fforootd** [reply]
Ok interesting approach. I need to think about this tommorow in more depth.

What I was thinking of is more or less this here https://doc.dovecot.org/configuration_manual/authentication/#authentication-in-proxies-and-directors
